server/neptune/temporalworker/job/stream_handler.go

```diff
@@ -29,78 +29,75 @@ func NewTestStreamHandler(
 	}
 
 	return &StreamHandler{
-		JobOutput:        streamChan,
 		Store:            jobStore,
 		ReceiverRegistry: receiverRegistry,
 		LogFilter:        logFilter,
 		Logger:           logger,
 	}
 }
 
-func NewStreamHandler(
-	jobStore Store,
-	receiverRegistry ReceiverRegistry,
-	logFilters valid.TerraformLogFilters,
-	logger logging.Logger,
-) (*StreamHandler, StreamCloserFn) {
-
-	logFilter := filter.LogFilter{
-		Regexes: logFilters.Regexes,
-	}
-
-	streamChan := make(chan *OutputLine)
-	streamCloserFn := func() {
-		close(streamChan)
-	}
-
-	return &StreamHandler{
-		JobOutput:        streamChan,
-		Store:            jobStore,
-		ReceiverRegistry: receiverRegistry,
-		LogFilter:        logFilter,
-		Logger:           logger,
-	}, streamCloserFn
-}
-
 type StreamHandler struct {
-	JobOutput        chan *OutputLine
 	Store            Store
+	wg               sync.WaitGroup
 	ReceiverRegistry ReceiverRegistry
 	LogFilter        filter.LogFilter
 	Logger           logging.Logger
 }
 
-func (s *StreamHandler) Stream(jobID string, msg string) {
-	s.JobOutput <- &OutputLine{
-		JobID: jobID,
-		Line:  msg,
-	}
-}
-
-func (s *StreamHandler) Handle() {
-	for msg := range s.JobOutput {
-		// Filter out log lines from job output
-		if s.LogFilter.ShouldFilterLine(msg.Line) {
-			continue
+// RegisterJob returns a channel and creates a receiver go routine
+// that reads from the associated channel.
+//
+// New resources are created each time this is called so it is up to the consumer to ensure
+// this is not being called more than once per job id.
+//
+// Note: Cleanup will be blocked until these routines are complete, therefore
+// it is essential these channels are closed or we risk deadlock on shutdown
+func (s *StreamHandler) RegisterJob(id string) chan string {
```

Comment:
One thing to note is that since we use the same job id for `TerraformInit` and `TerraformPlan` activities, this will likely get called twice, first for init and then for plan. However, since the jobs are sequential, the channel created for init should get cleaned up when init activity completes and a new channel will be created for plan with the same job id. I doubt this would be a problem unless both both the  channels are open at the same time which is not the case here.

Reply:
even if both channels are open at the same time this wouldn't matter, we'd have two separate go routines reading from them.